### ISSUE TITLE
Build failure with spirv-llvm-translator 15.x

### ISSUE THREAD

**Issue by ArchangeGabriel** [ispc/ispc#2419]
While trying to upgrade our stack to LLVM 15, we faced a build failure in ispc git head:
```cxx
/build/ispc/src/ispc-1.18.1.r285+ga77f4e2/src/opt/MangleOpenCLBuiltins.cpp:66:5: error: no matching function for call to 'mangleOpenClBuiltin'
    mangleOpenClBuiltin(funcName, ArgTy, mangledName);
    ^~~~~~~~~~~~~~~~~~~
/usr/include/LLVMSPIRVLib/LLVMSPIRVLib.h:166:6: note: candidate function not viable: requires 4 arguments, but 3 were provided
void mangleOpenClBuiltin(const std::string &UnmangledName,
     ^
/build/ispc/src/ispc-1.18.1.r285+ga77f4e2/src/opt/MangleOpenCLBuiltins.cpp:73:5: error: no matching function for call to 'mangleOpenClBuiltin'
    mangleOpenClBuiltin(func.getName().str(), func.getArg(0)->getType(), mangledName);
    ^~~~~~~~~~~~~~~~~~~
/usr/include/LLVMSPIRVLib/LLVMSPIRVLib.h:166:6: note: candidate function not viable: requires 4 arguments, but 3 were provided
void mangleOpenClBuiltin(const std::string &UnmangledName,
     ^
/build/ispc/src/ispc-1.18.1.r285+ga77f4e2/src/opt/MangleOpenCLBuiltins.cpp:87:5: error: no matching function for call to 'mangleOpenClBuiltin'
    mangleOpenClBuiltin(func.getName().str(), func.getArg(0)->getType(), mangledName);
    ^~~~~~~~~~~~~~~~~~~
/usr/include/LLVMSPIRVLib/LLVMSPIRVLib.h:166:6: note: candidate function not viable: requires 4 arguments, but 3 were provided
void mangleOpenClBuiltin(const std::string &UnmangledName,
     ^
```
This is due to a change in spirv-llvm-translator that is part of the release targeting LLVM 15: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/0eb9a7d2937542e1f95a4e1f9aa9850e669dc45f.

**Comment by foutrelis** [comment]
It also seems that this will be changed again for LLVM 16 because of https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/42cf770344bb8d0a32db1ec892bee63f43d793b1 which is only on the `main` branch currently.

**Comment by aneshlya** [comment]
ISPC with Xe targets doesn't support the build with LLVM 15 and newer yet. The recommended and the most tested version is LLVM 14.0.

**Comment by foutrelis** [comment]
If the [following](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/0eb9a7d2937542e1f95a4e1f9aa9850e669dc45f#diff-e3a504a02e520407f4cc40de54fe430a16a21dd92f2ee7a7ae953eb22158214cR165) is true for ISPC's calls to `mangleOpenClBuiltin` (not sure if it is!):

```
If there are no pointer-typed arguments in ArgTypes, then ArgPointerTypes may be empty.
```

Could we possibly use the following to unblock LLVM 15's roll-out to Arch Linux? (Seems to pass most tests.)

```diff
diff -upr ispc-1.18.1.orig/src/opt/MangleOpenCLBuiltins.cpp ispc-1.18.1/src/opt/MangleOpenCLBuiltins.cpp
--- ispc-1.18.1.orig/src/opt/MangleOpenCLBuiltins.cpp	2023-01-10 20:57:49.366931249 +0000
+++ ispc-1.18.1/src/opt/MangleOpenCLBuiltins.cpp	2023-01-10 21:05:14.169089807 +0000
@@ -45,6 +45,7 @@ static std::string mangleMathOCLBuiltin(
     llvm::Type *retType = func.getReturnType();
     std::string funcName = func.getName().str();
     std::vector<llvm::Type *> ArgTy;
+    std::vector<SPIRV::PointerIndirectPair> ArgPointerTy;
     // spirv OpenCL builtins are used for double types only
     Assert(retType->isVectorTy() && llvm::dyn_cast<llvm::FixedVectorType>(retType)->getElementType()->isDoubleTy() ||
            retType->isSingleValueType() && retType->isDoubleTy());
@@ -63,14 +64,15 @@ static std::string mangleMathOCLBuiltin(
     } else if (retType->isSingleValueType() && retType->isDoubleTy()) {
         ArgTy.push_back(LLVMTypes::DoubleType);
     }
-    mangleOpenClBuiltin(funcName, ArgTy, mangledName);
+    mangleOpenClBuiltin(funcName, ArgTy, ArgPointerTy, mangledName);
     return mangledName;
 }
 
 static std::string manglePrintfOCLBuiltin(const llvm::Function &func) {
     Assert(func.getName() == "__spirv_ocl_printf" && "wrong argument: ocl builtin is expected");
     std::string mangledName;
-    mangleOpenClBuiltin(func.getName().str(), func.getArg(0)->getType(), mangledName);
+    std::vector<SPIRV::PointerIndirectPair> ArgPointerTy;
+    mangleOpenClBuiltin(func.getName().str(), func.getArg(0)->getType(), ArgPointerTy, mangledName);
     return mangledName;
 }
 
@@ -84,7 +86,8 @@ static std::string mangleOCLBuiltin(cons
 static std::string mangleSPIRVBuiltin(const llvm::Function &func) {
     Assert(func.getName().startswith("__spirv_") && "wrong argument: spirv builtin is expected");
     std::string mangledName;
-    mangleOpenClBuiltin(func.getName().str(), func.getArg(0)->getType(), mangledName);
+    std::vector<SPIRV::PointerIndirectPair> ArgPointerTy;
+    mangleOpenClBuiltin(func.getName().str(), func.getArg(0)->getType(), ArgPointerTy, mangledName);
     return mangledName;
 }
 
```

**Comment by aneshlya** [comment]
> If the [following](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/0eb9a7d2937542e1f95a4e1f9aa9850e669dc45f#diff-e3a504a02e520407f4cc40de54fe430a16a21dd92f2ee7a7ae953eb22158214cR165) is true for ISPC's calls to `mangleOpenClBuiltin` (not sure if it is!):
> 
> ```
> If there are no pointer-typed arguments in ArgTypes, then ArgPointerTypes may be empty.
> ```
> 
> Could we possibly use the following to unblock LLVM 15's roll-out to Arch Linux? (Seems to pass most tests.)
> 
> ```diff
> diff -upr ispc-1.18.1.orig/src/opt/MangleOpenCLBuiltins.cpp ispc-1.18.1/src/opt/MangleOpenCLBuiltins.cpp
> --- ispc-1.18.1.orig/src/opt/MangleOpenCLBuiltins.cpp	2023-01-10 20:57:49.366931249 +0000
> +++ ispc-1.18.1/src/opt/MangleOpenCLBuiltins.cpp	2023-01-10 21:05:14.169089807 +0000
> @@ -45,6 +45,7 @@ static std::string mangleMathOCLBuiltin(
>      llvm::Type *retType = func.getReturnType();
>      std::string funcName = func.getName().str();
>      std::vector<llvm::Type *> ArgTy;
> +    std::vector<SPIRV::PointerIndirectPair> ArgPointerTy;
>      // spirv OpenCL builtins are used for double types only
>      Assert(retType->isVectorTy() && llvm::dyn_cast<llvm::FixedVectorType>(retType)->getElementType()->isDoubleTy() ||
>             retType->isSingleValueType() && retType->isDoubleTy());
> @@ -63,14 +64,15 @@ static std::string mangleMathOCLBuiltin(
>      } else if (retType->isSingleValueType() && retType->isDoubleTy()) {
>          ArgTy.push_back(LLVMTypes::DoubleType);
>      }
> -    mangleOpenClBuiltin(funcName, ArgTy, mangledName);
> +    mangleOpenClBuiltin(funcName, ArgTy, ArgPointerTy, mangledName);
>      return mangledName;
>  }
>  
>  static std::string manglePrintfOCLBuiltin(const llvm::Function &func) {
>      Assert(func.getName() == "__spirv_ocl_printf" && "wrong argument: ocl builtin is expected");
>      std::string mangledName;
> -    mangleOpenClBuiltin(func.getName().str(), func.getArg(0)->getType(), mangledName);
> +    std::vector<SPIRV::PointerIndirectPair> ArgPointerTy;
> +    mangleOpenClBuiltin(func.getName().str(), func.getArg(0)->getType(), ArgPointerTy, mangledName);
>      return mangledName;
>  }
>  
> @@ -84,7 +86,8 @@ static std::string mangleOCLBuiltin(cons
>  static std::string mangleSPIRVBuiltin(const llvm::Function &func) {
>      Assert(func.getName().startswith("__spirv_") && "wrong argument: spirv builtin is expected");
>      std::string mangledName;
> -    mangleOpenClBuiltin(func.getName().str(), func.getArg(0)->getType(), mangledName);
> +    std::vector<SPIRV::PointerIndirectPair> ArgPointerTy;
> +    mangleOpenClBuiltin(func.getName().str(), func.getArg(0)->getType(), ArgPointerTy, mangledName);
>      return mangledName;
>  }
>  
> ```

There can be pointer-typed arguments in ISPC. I'll fix the build, expect it the next couple of days. But again, even if build is successful it doesn't mean that ISPC is fully functional. Pleas let us know if building with LLVM 15 is critical and what blocks you from building with LLVM 14.

**Comment by foutrelis** [comment]
Many thanks for planning to address the build failure. We understand that there might still be issues during runtime.

As a rolling distro (Arch Linux) we try to build dependent packages against the latest version of LLVM. This is often not possible immediately after new LLVM releases which is why we delay updating by a few months. We do have separate LLVM 14 packages for more self-contained software to use (Crystal, Julia, LDC). However, building ISPC against it would be tricky because of the following dependency chains:

- spirv-llvm-translator 15.0.0 ← ispc
- spirv-llvm-translator 15.0.0  ← libclc 15.0.6 ← mesa
- spirv-llvm-translator 15.0.0 ← intel-opencl-clang 15.0.0 ← intel-graphics-compiler ← intel-compute-runtime

If we were to build ISPC against LLVM 14, I believe we would have to downgrade all of the above packages which currently track LLVM 15 (spirv-llvm-translator, intel-opencl-clang, libclc) as well as build Mesa against LLVM 14. Alternatively, ISPC (and IGC / compute-runtime?) could use a separate spirv-llvm-translator-14 package, but I am not sure how well this would work in practice.

**Comment by ArchangeGabriel** [comment]
Fixed by #2426.
